src/Microsoft.Diagnostics.EventFlow.Inputs.PerformanceCounter/TrackedPerformanceCounter.cs

```diff
@@ -0,0 +1,77 @@
+﻿// ------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All rights reserved.
+//  Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.Diagnostics.EventFlow.Configuration;
+using Validation;
+
+namespace Microsoft.Diagnostics.EventFlow.Inputs
+{
+    internal class TrackedPerformanceCounter : IDisposable
+    {
+        private PerformanceCounter counter;
+        private bool disposed;
+        private DateTimeOffset lastAccessedOn;
+        private string lastInstanceName;
+
+        public PerformanceCounterConfiguration Configuration { get; private set; }
+
+        public TrackedPerformanceCounter(PerformanceCounterConfiguration configuration)
+        {
+            Requires.NotNull(configuration, nameof(configuration));
+
+            this.Configuration = configuration;
+            this.lastAccessedOn = DateTimeOffset.UtcNow.Subtract(TimeSpan.FromDays(1));
+            this.disposed = false;
+        }
+
+        public bool SampleNextValue(ProcessInstanceNameCache processInstanceNameCache, out float newValue)
+        {
+            Requires.NotNull(processInstanceNameCache, nameof(processInstanceNameCache));
+
+            newValue = 0;
+            if (disposed)
+            {
+                return false;
+            }
+
+            var now = DateTimeOffset.UtcNow;
+            if (now - this.lastAccessedOn < TimeSpan.FromMilliseconds(this.Configuration.SamplingIntervalMsec))
+            {
+                return false;
+            }
+
+            string instanceName = processInstanceNameCache.GetCounterInstanceNameForCurrentProcess(this.Configuration);
+            this.lastAccessedOn = now;
+            if (this.counter == null || (!string.IsNullOrEmpty(instanceName) && instanceName != this.lastInstanceName))
```

Comment:
The PerformanceCounter instance naming scheme that is used by most counters is exename, exename#1, exename#2 and so on (assuming the interesting case, that is, multiple processes that are based on the same executable). But these instance names are not stable, meaning that as the processes start and stop, their instance names might change, because the system is trying to avoid "gaps" between instance names. Say you have two processes, first uses exename for counter instances, second uses exename#1. Then the first process ends. Now the second process will use exename for its instances.

It is kind of messy